Don't create scaling vector when we aren't doing AD

lindsayad opened the pull request:
With the merge of #16440, global AD became our default config. Griffin
revealed some undesirable behavior in the global AD code-base
post-merge. We were creating the `NumericVector` of scaling factors
anytime that any variable had a non-unity scaling factor; however, this
vector is actually only ever used through AD residual/jacobian computing
objects. There is another issue which is that the
`assembleScalingVector` code does not yet work with array variables,
which is the reason Griffin failed: we were trying to assemble this
unnecessary vector, and that code triggered the assertion about not
working with array variables. https://github.com/idaholab/moose/issues/16820 created for that.
In the meantime (and in the future) we will not build the scaling vector
if we do not have AD objects

lindsayad committed:
Don't create scaling vector when we aren't doing AD

With the merge of #16440, global AD became our default config. Griffin
revealed some undesirable behavior in the global AD code-base
post-merge. We were creating the `NumericVector` of scaling factors
anytime that any variable had a non-unity scaling factor; however, this
vector is actually only ever used through AD residual/jacobian computing
objects. There is another issue which is that the
`assembleScalingVector` code does not yet work with array variables,
which is the reason Griffin failed: we were trying to assemble this
unnecessary vector, and that code triggered the assertion about not
working with array variables. I will create an issue for that problem.
In the meantime (and in the future) we will not build the scaling vector
if we do not have AD objects

lindsayad commented:
@YaqiWang this patch fixed the Griffin assertions for me. I will add a simple MOOSE test here

YaqiWang commented:
Thanks Alex.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/669147/) on c33d7ad wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16819/site)

This comment will be updated on new commits.

GiudGiud reviewed (commented):
dug down the doxygen to see what all the setup that were called are and why but someone with more experience on the setup should review.

lindsayad committed:
Each problem must inform the other if AD objects have been added

lindsayad commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -3546,7 +3563,7 @@ NonlinearSystemBase::assembleScalingVector()
   {
     mooseAssert(scalar_var->count() == 1,
-                "Contact Robert Carlsen and tell him to make this work for array variables.");
+                "Contact Alex Lindsay and tell him to make this work for array variables.");
```
:smile:

lindsayad committed:
Always addScalingVector from MooseVariableBase::initialSetup

There's no performance penalty to do thing this because this calls
through to SystemBase::addVector which checks whether we already have
the vector and if so, just returns it. And now that `addScalingVector`
happens much later in the simulation (e.g. during `initialSetup` instead
of during variable construction) we need to make sure that we always add
the vector, because this triggers retrieval of the vector in assembly.
If we kept the `hasVector` check in `MooseVariableBaes::initialSetup` we
would actually never trigger the vector retrieval in Assembly because
`hasVector` is now `true` by the time we arrive at
`MooseVariableBase::initialSetup`. This is because we have restored the
libmesh systems and vectors by this point in the simulation
